x/ecocredit/server/msg_server.go

```diff
@@ -777,12 +777,24 @@ func (s serverImpl) Buy(goCtx context.Context, req *ecocredit.MsgBuy) (*ecocredi
 				return nil, err
 			}
 
+			// error if auto-retire is required for given sell order
+			if !sellOrder.DisableAutoRetire && order.DisableAutoRetire {
+				return nil, ecocredit.ErrInvalidBuyOrder.Wrapf("auto-retire required for sell order %d", sellOrder.OrderId)
+			}
+
+			// declare credit for send message
 			credit := &ecocredit.MsgSend_SendCredits{
-				BatchDenom:     sellOrder.BatchDenom,
-				TradableAmount: creditsToReceive.String(),
-				// TODO: handle auto-retire settings #621
-				RetiredAmount: "0",
-				//RetirementLocation: retirementLocation,
+				BatchDenom: sellOrder.BatchDenom,
+			}
+
+			// set tradable or retired amount depending on auto-retire
+			if sellOrder.DisableAutoRetire && order.DisableAutoRetire {
+				credit.RetiredAmount = "0"
+				credit.TradableAmount = creditsToReceive.String()
+			} else {
+				credit.RetiredAmount = creditsToReceive.String()
+				credit.RetirementLocation = order.RetirementLocation
+				credit.TradableAmount = "0"
```

Comment:
should we fail the buy order here in the case of a sell order with auto-retirement enabled and the buy order has it disabled? 

seems like the direction we wanted to go based on [comment](https://github.com/regen-network/regen-ledger/blob/ryan/621-buy-auto-retire/proto/regen/ecocredit/v1alpha2/types.proto#L202-L207)

Reply:
There's a condition on lines 780-783 that throws an error if enabled for the sell order and disabled for the buy order.

The current logic is as follows:
- disabled for sell order, disabled for buy order, successfully receive tradable credits
- enabled for sell order, enabled for buy order, successfully receive retired credits
- enabled for sell order, disabled for buy order, fail with error
- disabled for sell order, enabled for buy order, successfully receive retired credits

Reply:
There's also a test for this `name: "auto-retire required for sell order"`